### ISSUE TITLE
Some build times for single case are slightly longer in cime5

### ISSUE THREAD

**Issue by billsacks** [ESMCI/cime#312]
This is a minor issue, and if the cause is either unclear or hard to fix, I'm okay with this being deferred or labeled as a wontfix. But I wanted to document it since it seems worth trying to keep build times down as much as possible.

I have noticed that build times for single cases are slightly longer in cime5 than they were in cime4, with my main experience being with I compsets (CLM with datm) with CLM45 / CLM50 (which is built in the shared lib stage) on yellowstone. The difference is only about 30 sec (out of a 6-7 min total build time) - so not huge, but I'm not sure why there should be any difference.

I have done some relatively objective timings as follows ("relatively" objective because there could still have been machine variability from one build to the next, particularly because I built on a shared queue so did not have exclusive access to the node):

I got an interactive session on caldera with:

```
DAV_CORES=8 execca
```

First, with cime5, with these externals:

```
cime                               https://github.com/CESM-Development/cime/tags/cime5.0.3
components/cam                     https://svn-ccsm-models.cgd.ucar.edu/cam1/branch_tags/cam5_4_56_esmciport_tags/cam5_4_56_esmciport_n08/components/cam
components/clm                     https://svn-ccsm-models.cgd.ucar.edu/clm2/branch_tags/clm4_5_8_r182_esmciport_tags/clm4_5_8_r182_esmciport_n03/components/clm
components/cice                    https://svn-ccsm-models.cgd.ucar.edu/cice/branch_tags/cice5_20160401_esmciport_tags/cice5_20160401_esmciport_n02/
components/pop                     https://svn-ccsm-models.cgd.ucar.edu/pop2/branch_tags/cesm_pop_2_1_20160426_esmciport_tags/cesm_pop_2_1_20160426_esmciport_n03
components/cism                    https://svn-ccsm-models.cgd.ucar.edu/glc/branch_tags/esmci_cime_tags/esmci_cime_n01_cism2_1_18/
components/aquap                   https://svn-ccsm-models.cgd.ucar.edu/aquap/trunk_tags/aquap_151016
components/rtm                     https://svn-ccsm-models.cgd.ucar.edu/rivrtm/trunk_tags/rtm1_0_56
components/ww3                     https://svn-ccsm-models.cgd.ucar.edu/ww3/trunk_tags/ww3_160611
components/mosart                  https://svn-ccsm-models.cgd.ucar.edu/mosart/trunk_tags/mosart1_0_17
```

I timed just the build stage:

```
./create_test --no-build SMS.f10_f10.ICLM45.yellowstone_intel

time ./case.build
```

This gave:

```
Time spent not building: 22.865428 sec
Time spent building: 370.595688 sec

real 7m4.058s
user 9m46.932s
sys 4m13.146s
```

Then, from CLM r182, using cime4:

```
./create_test -testname SMS.f10_f10.ICLM45.yellowstone_intel

time ./case.test_build
```

This gave:

```
      Mon Aug  1 13:05:49 2016 /glade/scratch/sacks/SMS.f10_f10.ICLM45.yellowstone_intel.160801-125110/bld/intel/mpich2/nodebug/nothreads/mct.bldlog.160801-130523
      Mon Aug  1 13:07:37 2016 /glade/scratch/sacks/SMS.f10_f10.ICLM45.yellowstone_intel.160801-125110/bld/intel/mpich2/nodebug/nothreads/gptl.bldlog.160801-130523
      Mon Aug  1 13:07:49 2016 /glade/scratch/sacks/SMS.f10_f10.ICLM45.yellowstone_intel.160801-125110/bld/intel/mpich2/nodebug/nothreads/pio.bldlog.160801-130523
      Mon Aug  1 13:08:33 2016 /glade/scratch/sacks/SMS.f10_f10.ICLM45.yellowstone_intel.160801-125110/bld/intel/mpich2/nodebug/nothreads/csm_share.bldlog.160801-130523
    .... calling builds for component libraries
model = atm, obj = /glade/scratch/sacks/SMS.f10_f10.ICLM45.yellowstone_intel.160801-125110/bld/atm/obj
      .... calling /glade/scratch/sacks/cesm_code/clm4_5_8_r182/cime/components/data_comps/datm/cime_config/buildlib
           Mon Aug  1 13:09:11 2016 /glade/scratch/sacks/SMS.f10_f10.ICLM45.yellowstone_intel.160801-125110/bld/atm.bldlog.160801-130523
         - Building clm4_5/clm5_0 shared library
      .... calling /glade/scratch/sacks/cesm_code/clm4_5_8_r182/components/clm/cime_config/buildlib
           Mon Aug  1 13:09:16 2016 /glade/scratch/sacks/SMS.f10_f10.ICLM45.yellowstone_intel.160801-125110/bld/lnd.bldlog.160801-130523
model = ice, obj = /glade/scratch/sacks/SMS.f10_f10.ICLM45.yellowstone_intel.160801-125110/bld/ice/obj
      .... calling /glade/scratch/sacks/cesm_code/clm4_5_8_r182/cime/components/stub_comps/sice/cime_config/buildlib
           Mon Aug  1 13:11:17 2016 /glade/scratch/sacks/SMS.f10_f10.ICLM45.yellowstone_intel.160801-125110/bld/ice.bldlog.160801-130523
model = ocn, obj = /glade/scratch/sacks/SMS.f10_f10.ICLM45.yellowstone_intel.160801-125110/bld/ocn/obj
      .... calling /glade/scratch/sacks/cesm_code/clm4_5_8_r182/cime/components/stub_comps/socn/cime_config/buildlib
           Mon Aug  1 13:11:19 2016 /glade/scratch/sacks/SMS.f10_f10.ICLM45.yellowstone_intel.160801-125110/bld/ocn.bldlog.160801-130523
model = glc, obj = /glade/scratch/sacks/SMS.f10_f10.ICLM45.yellowstone_intel.160801-125110/bld/glc/obj
      .... calling /glade/scratch/sacks/cesm_code/clm4_5_8_r182/cime/components/stub_comps/sglc/cime_config/buildlib
           Mon Aug  1 13:11:20 2016 /glade/scratch/sacks/SMS.f10_f10.ICLM45.yellowstone_intel.160801-125110/bld/glc.bldlog.160801-130523
model = wav, obj = /glade/scratch/sacks/SMS.f10_f10.ICLM45.yellowstone_intel.160801-125110/bld/wav/obj
      .... calling /glade/scratch/sacks/cesm_code/clm4_5_8_r182/cime/components/stub_comps/swav/cime_config/buildlib
           Mon Aug  1 13:11:22 2016 /glade/scratch/sacks/SMS.f10_f10.ICLM45.yellowstone_intel.160801-125110/bld/wav.bldlog.160801-130523
model = rof, obj = /glade/scratch/sacks/SMS.f10_f10.ICLM45.yellowstone_intel.160801-125110/bld/rof/obj
      .... calling /glade/scratch/sacks/cesm_code/clm4_5_8_r182/components/rtm/cime_config/buildlib
           Mon Aug  1 13:11:24 2016 /glade/scratch/sacks/SMS.f10_f10.ICLM45.yellowstone_intel.160801-125110/bld/rof.bldlog.160801-130523
model = esp, obj = /glade/scratch/sacks/SMS.f10_f10.ICLM45.yellowstone_intel.160801-125110/bld/esp/obj
      .... calling /glade/scratch/sacks/cesm_code/clm4_5_8_r182/cime/components/stub_comps/sesp/cime_config/buildlib
           Mon Aug  1 13:11:38 2016 /glade/scratch/sacks/SMS.f10_f10.ICLM45.yellowstone_intel.160801-125110/bld/esp.bldlog.160801-130523
      Mon Aug  1 13:11:39 2016 /glade/scratch/sacks/SMS.f10_f10.ICLM45.yellowstone_intel.160801-125110/bld/cesm.bldlog.160801-130523
 .... locking file ./env_build.xml

real 6m42.334s
user 9m49.371s
sys 3m27.482s

-rw-r--r-- 1 sacks ncar     1835 01-08-2016 13:11:55 cesm.bldlog.160801-130523.gz
-rwxr-xr-x 1 sacks ncar 49775266 01-08-2016 13:12:04 cesm.exe
-rwxr-xr-x 1 sacks ncar 49775266 01-08-2016 13:12:04 cesm.exe.160801-130523
```

So this took slightly less than the 7:04 in cime5. Also, for comparison with the following timing, note that it took about 6:15 from the start of the actual build to the production of cesm.exe.

Finally, back in cime5, I looked at the timing of running create_test without --no-build:

```
time ./create_test SMS.f10_f10.ICLM45.yellowstone_intel

bash-4.1$ time ./create_test SMS.f10_f10.ICLM45.yellowstone_intel
Guessing CIME_MODEL=cesm, set environment variable if this is incorrect
Using project from .cesm_proj: P93300601
RUNNING TESTS:
  SMS.f10_f10.ICLM45.yellowstone_intel
Starting CREATE_NEWCASE for test SMS.f10_f10.ICLM45.yellowstone_intel with 1 procs
Finished CREATE_NEWCASE for test SMS.f10_f10.ICLM45.yellowstone_intel in 2.251607 seconds (PASS)
Starting XML for test SMS.f10_f10.ICLM45.yellowstone_intel with 1 procs
Finished XML for test SMS.f10_f10.ICLM45.yellowstone_intel in 0.263287 seconds (PASS)
Starting SETUP for test SMS.f10_f10.ICLM45.yellowstone_intel with 1 procs
Finished SETUP for test SMS.f10_f10.ICLM45.yellowstone_intel in 20.211938 seconds (PASS)
Starting SHAREDLIB_BUILD for test SMS.f10_f10.ICLM45.yellowstone_intel with 1 procs
Finished SHAREDLIB_BUILD for test SMS.f10_f10.ICLM45.yellowstone_intel in 333.997401 seconds (PASS)
Starting MODEL_BUILD for test SMS.f10_f10.ICLM45.yellowstone_intel with 4 procs
Finished MODEL_BUILD for test SMS.f10_f10.ICLM45.yellowstone_intel in 77.529663 seconds (PASS)
```

So SHAREDLIB_BUILD + MODEL_BUILD took about 6:52 together -- about 37 sec longer than the time from the start of the actual build to the production of cesm.exe in cime4.

It's possible that these comparisons aren't completely fair, because I needed to use different methods to determine the timing of each version (and, for cime5, different methods depending on whether I manually ran case.build or ran the build as part of create_test).

**Comment by jgfouca** [comment]
That's kinda weird because CIME5 parallelizes component builds.

**Comment by jedwards4b** [comment]
This is a particular issue for clm5 which is built with the shared_libraries and since I cases are clm only almost all of the time is in the serialized shared build.

**Comment by billsacks** [comment]
I have noticed that the time spent just in the python code in the SHAREDLIB_BUILD portion, before even getting to build itself (i.e., the makefile portion of the build) can take > 10 sec. (I noticed this in a test that failed during the testcase's build step, in the constructor for the test case.) That might be enough to explain what I'm seeing. 

Optimization of the python scripts could be worthwhile, but it sounds like that is already on someone's radar (@jgfouca, I think). So I'm going to close this issue.
